cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java

```diff
@@ -148,6 +148,143 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
+
+    List<String> metricsScanResults = Lists.newArrayList();
+    Row rowResult;
+    int contextOffset = entityCodec.getEncodedSize(MetricsEntityType.CONTEXT);
+
+
+    // multiple scans with incrementing the scans startRow row-key to get the next unique part of a context or
+    // next unique metric based on the parameter isContextQuery.
+    // by this way, we can skip multiple rows of duplicate rows and scan for the next
+    // unique context/metric part we are interested in.
+    // we stop when we cannot find any rows which matches the given contextPrefix or if there are no rows
+    // returned from the scan.
+    do {
+      ScannerFields fields = new ScannerFields(startRow, endRow, null, filter);
+      Scanner scanner = null;
+      try {
+        scanner = timeSeriesTable.scan(fields.startRow, fields.endRow, fields.columns, fields.filter);
+      } catch (Exception e) {
+        throw new OperationException(StatusCode.INTERNAL_ERROR, e.getMessage(), e);
+      }
+
+      rowResult = scanner.next();
+      if (rowResult != null) {
+        byte[] rowKey = rowResult.getRow();
+        String contextStr = entityCodec.decode(MetricsEntityType.CONTEXT, rowKey, 0);
+        if (contextPrefix != null && !contextStr.startsWith(contextPrefix)) {
+          // if retrieved rowkey's contextPrefix does not match with the contextPrefix in query,
+          // we stop scanning and return
+          scanner.close();
+          break;
+        }
+
+        if (isContextQuery) {
+          metricsScanResults.add(contextStr);
+        } else {
+          metricsScanResults.add(entityCodec.decode(MetricsEntityType.METRIC, rowKey, contextOffset));
+        }
+        // With the next scan we fast-forward to the next row that has different context/metric name
+        // that we are searching for.
+        startRow = getNextRow(rowKey, targetOffset, length);
+        if (startRow == null) {
+          //reached max possible key for the context, we will stop scanning now and return.
+          scanner.close();
+          break;
+        }
+      }
+      // no more matching rows found
+      scanner.close();
+    } while (rowResult != null);
+
+    return metricsScanResults;
+  }
+
+  /*
+   * Returns row key that is:
+   * greater than given rowKey AND
+   * has same bytes in the beginning (up to the given offset) AND
+   * has different bytes in the range [offset..offset+length]
+   * Returns null if no row key with this properties exist.
+   */
+  private byte[] getNextRow(byte[] rowKey, int offset, int length) {
+    byte[] stopKey = Bytes.stopKeyForPrefix(Arrays.copyOfRange(rowKey,
+                                                               offset, offset + length));
+    if (stopKey == null) {
+      return null;
+    }
+    // next row-key
+    byte[] nextRow = new byte[rowKey.length];
+
+    copyByteArray(rowKey, nextRow, 0, 0 , offset);
```

Comment:
a) allocate array of known size
b) use System.arrayCopy to fill prefix (that is unchanged)
c) use System.arrayCopy to fill stopKey in the rest